### ISSUE TITLE
Question on HiFi metagenome in V2.1

### ISSUE THREAD

**Issue by proteinosome** [marbl/canu#1796]
Hi, first of all thanks for developing Canu. I've been testing Canu 2.1 in the PacBio HiFi metagenome mock community data and is seeing quite a big improvement in terms of generating less contigs and getting a lot more circular contigs. I have a few observations/questions:

1. Is `'batOptions=-eg 0.0 -sb 0.001 -dg 0 -db 3 -dr 0 -ca 2000 -cp 200'` still recommended? Previously, I observed that in Canu 2.0, this produced more contigs but is more "correct" in the sense that there's more circular contigs and there's no crazy long knot (10Mbp+). However, in Canu 2.1, it appears that even without this option it assembles into lesser contigs and no longer has the long knot issue.

2. For metagenomes, `maxInputCoverage` seems to still be needed. Setting it to 1000 (everything else default) results in more "complete" genomes (higher total genome size), yet has more contigs > 1Mbp and higher N50. Is this expected?

3. In addition, I tried adding `'corOutCoverage=10000' 'corMhapSensitivity=high' 'corMinCoverage=0'` as well as with 1000 `maxInputCoverage`, and was able to get almost 2 fold decrease in number of contigs but similar total genome size and the number of circular contigs almost doubled as well. Separately, I tried removing either `'corMhapSensitivity=high'` or `'corOutCoverage=10000'` and both don't seem to have an effect.

4. Which brings me to the question, which of the `corX` parameters actually affect HiFi assembly since the correction step is quite different? From 3, it seems that `'corMinCoverage=0'` has an important (and positive) effect on assembly?

5. In all cases, I filtered the contigs marked as bubbles. I saw this being recommended in some of the previous issues. Is this still true in 2.1 and do you think the bubble may be potentially useful in separating strains? They add a lot more small contigs to the final assembly.

Thank you!

**Comment by skoren** [comment]
1. Those options are still a bit more conservative than the defaults but the defaults should work pretty well in 2.1.

2. Yes, maxInputCoverage defaults to 50x for hifi data so metagenomes definitely need to overwrite that.

3/4. None of the cor options will affect HiFi assembly as it doesn't do correction or trimming. My guess is you're still subsetting the coverage even at maxInputCoverage=1000 and that's leading to different results. I don't see how corMinCoverage would affect HiFi results otherwise.

5. The bubbles are flagged based on what is likely to cause false confusion and repeat splitting, so they are certainly not going to be comprehensive. They likely will catch some variation between similar genomes but they will not catch larger-scale variation. For that reason, by filtering bubbles, you may be only filtering part of the variation in the dataset. In normal diploid assemblies, we usually include bubbles and rely on another tool like purge-dups to select a single haplotype. However, that's not an option in a metagenomic dataset. As a first pass, it's probably OK to not include bubbles. I'd still look at them at some point though because they are likely real sequence.

**Comment by proteinosome** [comment]
Hi Sergey,

Thank you so much for your reply! For 3/4, in all the runs where I kept `corMinCoverage=0` in addition to `maxInputCoverage=1000`, the results are identical (205 contigs, 55Mbp genome size, 29 circular contigs), so this likely rules out sub-sampling. In a run where I only added `maxInputCoverage=1000` without the `cor` options, the results are 393 contigs, 55Mbp genome size, 17 circular contigs.

I noticed in the report, unitigging error rate is different comparing with and without the `cor` options. I've attached the run scripts and run reports here for your reference.

Default parameter with `'maxInputCoverage=1000'` (n=393):
[canu.01.cov1000.sh.txt](https://github.com/marbl/canu/files/5229989/canu.01.cov1000.sh.txt)
[ATCC-MSA1003-canu-2.1-cov1000.report.txt](https://github.com/marbl/canu/files/5229985/ATCC-MSA1003-canu-2.1-cov1000.report.txt)

With `'maxInputCoverage=1000' 'corMhapSensitivity=high' 'corMinCoverage=0' 'redMemory=32' 'oeaMemory=32' 'batMemory=200'` (n=205):
[canu.01.corMhap_cov1000.sh.txt](https://github.com/marbl/canu/files/5229988/canu.01.corMhap_cov1000.sh.txt)
[ATCC-MSA1003-canu-2.1_corMhap_cov1000.report.txt](https://github.com/marbl/canu/files/5229987/ATCC-MSA1003-canu-2.1_corMhap_cov1000.report.txt)

With `'maxInputCoverage=1000' 'corOutCoverage=10000' 'corMhapSensitivity=high' 'corMinCoverage=0' 'redMemory=32' 'oeaMemory=32' 'batMemory=200'` (n=205):
[canu.01.custom-cor-cov1000.sh.txt](https://github.com/marbl/canu/files/5229990/canu.01.custom-cor-cov1000.sh.txt)
[ATCC-MSA1003-canu-2.1-custom-cor-cov1000.report.txt](https://github.com/marbl/canu/files/5229986/ATCC-MSA1003-canu-2.1-custom-cor-cov1000.report.txt)

Rather puzzled by this as well...

**Comment by brianwalenz** [comment]
Jsut to check, you are running with `-pacbio-hifi`, correct?  I think so based on the report files, as they don't mention anything about correction or trimming (and this is the correct behavior for pacbio-hifi).

Compare 4-unitigger/unitigger.err between the n=393 and n=205 runs.  If you have A LOT of overlaps, the n=393 run could be discarding shorter overlaps to reach the default memory limit of 16-64 GB (the specific value is based on machine size).  You're looking for something like
```
OverlapCache()-- Loading overlaps.
OverlapCache()--
OverlapCache()--          read from store           saved in cache
OverlapCache()--   ------------ ---------   ------------ ---------
[probably lines here]
OverlapCache()--   ------------ ---------   ------------ ---------
OverlapCache()--        1022512 (100.00%)        1022487 (100.00%)    <---- expecting this to be < 100
```
There is some evidence this is happening in the report files.
```
--  BEST EDGE FILTERING
--  -------------------
--  At graph threshold 0.0300%, reads:
--    available to have edges:       378597
--    with at least one edge:        194764
```
vs
```
--  BEST EDGE FILTERING
--  -------------------
--  At graph threshold 0.0300%, reads:
--    available to have edges:       361191
--    with at least one edge:        337040
```
so the one without batMemory has fewer reads with 'at least one edge' - meaning there's around 120,000 reads with no useful overlap.

A "quick" test would be to run the n=393 unitigging stage by hand.  Make a directory parallel to 4-unitigger, copy in unitigger.sh, modify the script to include "-M 200", and then run the script.  The report file won't be updated, but you can diff some of the *.log files between the two runs.  The first *.log file (in ls -l order) has the error rate info.

**Comment by proteinosome** [comment]
Hi Brian, yes these are pacbio-hifi runs. Thanks I will try to run the test you suggested and get back to you. 

So in essence it's the batMemory options that's affecting the results? I.e. is setting higher batMemory important for HiFi metagenomic run?

Thank you!

**Comment by proteinosome** [comment]
Sorry for the late reply. I couldn't rerun the unitigging step as I deleted the `ovlStore` folder to save some space. I've tried adding `batMemory=200` option in addition to `maxInputCoverage=1000`, and the resulting contigs is 205, confirming that the `cor` options indeed did not do anything but it's the `batMemory` option that has an effect. Looking at the unitigging logs there's indeed a lot more overlaps compared to without the `batMemory` option.

In that case, will you always recommend using `batMemory` for HiFi metagenomics?
